contracts/PartyBuy.sol

```diff
@@ -0,0 +1,170 @@
+
+    function initialize(
+        address _nftContract,
+        uint256 _tokenId,
+        uint256 _maxPrice,
+        uint256 _secondsToTimeout,
+        Structs.AddressAndAmount calldata _split,
+        Structs.AddressAndAmount calldata _tokenGate,
+        string memory _name,
+        string memory _symbol
+    ) external initializer {
+        // validate maxPrice
+        require(_maxPrice > 0, "PartyBuy::initialize: must set price higher than 0");
+        // initialize & validate shared Party variables
+        __Party_init(_nftContract, _tokenId, _split, _tokenGate, _name, _symbol);
+        // set PartyBuy-specific state variables
+        expiresAt = block.timestamp + _secondsToTimeout;
+        maxPrice = _maxPrice;
+    }
+
+    // ======== External: Contribute =========
+
+    /**
+     * @notice Contribute to the Party's treasury
+     * while the Party is still active
+     * @dev Emits a Contributed event upon success; callable by anyone
+     */
+    function contribute() external payable nonReentrant {
+        // require that the new total contributed is not greater than
+        // the maximum amount the Party is willing to spend
+        require(totalContributedToParty + msg.value <= getMaximumContributions(), "PartyBuy::contribute: cannot contribute more than max");
+        // continue with shared _contribute flow
+        _contribute();
+    }
+
+    // ======== External: Buy =========
+
+    /**
+     * @notice Buy the token by calling targetContract with calldata supplying value
+     * @dev Emits a Bought event upon success; reverts otherwise. callable by anyone
+     */
+    function buy(uint256 _value, address _targetContract, bytes calldata _calldata) external nonReentrant {
+        require(
+            partyStatus == PartyStatus.ACTIVE,
+            "PartyBuy::buy: party not active"
+        );
+        // ensure the target contract is on allow list
+        require(allowList.allowed(_targetContract), "PartyBuy::buy: targetContract not on AllowList");
+        // check that value is not zero (else, token will be burned in TokenVault)
+        require(_value > 0, "PartyBuy::buy: can't spend zero");
+        // check that value is not more than the maximum price set at deploy time
+        require(_value <= maxPrice, "PartyBuy::buy: can't spend over max price");
+        // check that value is not more than
+        // the maximum amount the party can spend while paying ETH fee
+        require(_value <= getMaximumSpend(), "PartyBuy::buy: insuffucient funds to buy token plus fee");
+        // execute the calldata on the target contract
+        (bool _success, bytes memory _returnData) = address(_targetContract).call{value: _value}(_calldata);
+        // require that the external call succeeded
+        require(_success, string(_returnData));
+        // require that the NFT is owned by the Party
+        require(_getOwner() == address(this), "PartyBuy::buy: failed to buy token");
```

Comment:
yeah somebody sniping the NFT at a lower price, or sniping the NFT and devaluing it like this, is basically the major security-related question around this contract. 

the agreement with any `PartyBuy` is essentially "when you start this party, you are willing to spend any ETH sent into this party, as long as in the end you own NFT (A)." 
- the NFT could drop in price, you would still have agreed to spend any ETH in the party. 
- the NFT could be altered or devalued in some way, you would still have agreed to spend any ETH in the party.
i think the main issue, as with many nuanced security things related to Party, is that it's unclear if users deeply understand this.  @johncpalmer do we have anything in the UI to make this more clear?

this dynamic is we have a `maxPrice` cap - to prevent folks from depositing more ETH than the Party has agreed to pay.

Reply:
I'll note that this is why we have a `maxPrice` cap - to cap the amount of ETH possible to spend on the NFT, and prevent users from losing money by depositing too much 

this is also why we target one specific NFT - rather than, for example, any NFT from a given token contract; so that folks can determine the desired valuation of the NFT at deploy time

Reply:
> checking that the owner is not this contract before the call

this is when we would get into the deadlock described above where you cannot `buy` and you cannot `expire`

one could remove the requirement in `expire` that the NFT is not in the contract, but that would just open the potential to lock the NFT in the Party. a weird attack, but a possible one

Reply:
> > checking that the owner is not this contract before the call
> 
> this is when we would get into the deadlock described above where you cannot `buy` and you cannot `expire`
> 
> one could remove the requirement in `expire` that the NFT is not in the contract, but that would just open the potential to lock the NFT in the Party

Yeah but in this scenario the party never bought the NFT, it was placed there to grief the contract, so technically the Party was never entitled to it in the first place. If the attacker was willing to lose the NFT in this way, the effect of trapping funds is probably more worthwhile than the NFT itself, so maybe the priority should be on being able to gracefully return the funds as if the sale was lost. Removing the owner check in `expire()` sounds like the more correct logic in this context because it's more true to the intent/purpose of this contract?

Reply:
okay, I think this is good feedback, i removed the check from `expire` and I added a check before the external call in `buy`

now, if someone sends the NFT into the contract, the recourse is to `expire` the Party and users get their money back. the PartyDAO multisig may or may not also withdraw the NFT from the contract.

Reply:
i will note that profiting from altering / devaluing the NFT is still possible

1. Someone snipes the NFT (A).
2. They alter/devalue A somehow (pocketing some $ along the way).
4. They create another sale for NFT (A).
5. They call `buy` passing in _calldata to buy the new listing of A.
7. Attacker gets profit from devaluing A.

however, this is permissible within the context of the "agreement" of creating a PartyBuy.
 "when you start this party, you are willing to spend any ETH sent into this party, as long as in the end you own NFT (A)." - we don't guarantee anything about the state of NFT A, just its ultimate ownership

Reply:
Yeah that's a fair point. Guess there really is no getting around that attack.

Reply:
Although, in theory, without the owner check before the call, and without a selector filter, the second sale could be replaced with something else that routes ETH to the attacker, maybe more efficiently than a resale. Narrowing the range of possibilities is probably prudent with arbitrary external calls.